### PR TITLE
Make security copies of the key in RealmConfiguration

### PR DESCRIPTION
**Pull request by emanuelez**
Saving and providing a mutable reference of the key is both a security threat and possibly the origin of subtle bugs.
We now make security copies of the byte arrays to solve the issue.

### CODE REVIEW

**Comment by emanuelez** [comment]
@realm/java

**Comment by emanuelez** [comment]
Changelog: Fixed a bug which allowed to externally modify the encryption key in a RealmConfiguration

**Comment by beeender** [comment]
:+1:

**Comment by kneth** [comment]
:+1:

**Comment by cmelchior** [comment]
Shouldn't you do the copy when calling `encryptionKey(key)` as well?

**Comment by emanuelez** [comment]
@cm That's the second one (line 299)

**Comment by cmelchior** [comment]
Missing a unit test for it, but otherwise :+1:

**Comment by emanuelez** [comment]
retest this please

File: realm/src/androidTest/java/io/realm/RealmConfigurationTest.java [hunk 1]
```diff
@@ -416,4 +417,20 @@ public void testDifferentDirSameName() {
...
+    public void testKeyStorage() throws Exception {
+        // Generate a key and use it in a RealmConfiguration
+        byte[] oldKey = TestHelper.getRandomKey(12345);
+        byte[] key = oldKey;
+        RealmConfiguration config = new RealmConfiguration.Builder(getContext()).encryptionKey(key).build();
+
+        // Generate a different key and assign it to the same variable
+        byte[] newKey = TestHelper.getRandomKey(67890);
+        MoreAsserts.assertNotEqual(key, newKey);
+        key = newKey;
```
Comment (cmelchior) [inline realm/src/androidTest/java/io/realm/RealmConfigurationTest.java]: isn't the problem more if you do `key[0] = 42`, i.e changing the value inside the array?

**Pull request merged**